### ISSUE TITLE
Support secondary indexes on Apache Cassandra tables in YugaByte database

### ISSUE THREAD

**Issue by robertpang** [YugaByte/yugabyte-db#21]
Provide support for secondary indexes on Apache Cassandra tables in YugaByte database. The indexes should be strongly consistent, and can be global or local.

**Comment by shreklei** [comment]
When will the secondary index be released? Can I try it now?

**Comment by kmuthukk** [comment]
@shreklei : secondary indexes feature is in alpha. 

Currently, covered secondary indexes work; and non-covered secondary indexes are still work in progress. Secondary indexes in YugaByte are strongly consistent, and index updates use distributed transactions (across the primary table and the index).

One of us (likely @robertpang) will post a sample working example here soon.

**Comment by robertpang** [comment]
@shreklei Here is an example of using secondary index:

```SQL
cqlsh> create keyspace k;
cqlsh> use k;

-- Create primary table with "customer_id" as the partition key.
cqlsh:k> create table orders (
     ...   customer_id int,
     ...   order_date timestamp,
     ...   amount double,
     ...   primary key ((customer_id), order_date)
     ... ) with transactions = { 'enabled' : true };

-- Create secondary index on "orders" with "order_date" as the partition key.
cqlsh:k> create index orders_by_date on orders (order_date, customer_id) covering (amount);

-- Insert data.
cqlsh:k> insert into orders (customer_id, order_date, amount)
     ... values (1, '2018-04-02', 100.30);
cqlsh:k> insert into orders (customer_id, order_date, amount)
     ... values (2, '2018-04-02', 50.45);
cqlsh:k> insert into orders (customer_id, order_date, amount)
     ... values (1, '2018-04-06', 20.25);
cqlsh:k> insert into orders (customer_id, order_date, amount)
     ... values (3, '2018-04-06', 200.80);

-- Perform sum of order amount for a customer. This query will be executed against
-- the primary table using the partition key "customer_id".
cqlsh:k> select sum(amount) from orders where customer_id = 1;
 sum(amount)
-------------
      120.55
(1 rows)

-- Perform sum of order amount for a specific date. With the presence of the secondary
-- index, the query analyzer will execute the query against the secondary index using
-- the partition key "order_date" and avoid a full-table scan of the primary table.
cqlsh:k> select sum(amount) from orders where order_date = '2018-04-02';
 sum(amount)
-------------
      150.75
(1 rows)
```

A few additional points to note:
- To create a secondary index on a table, the primary table needs to be created with distributed transaction enabled using the `with transactions = { 'enabled' : true }` clause.
- The secondary index needs to be created before any data is inserted into the primary table.
- A secondary index will be used to execute a query only when the index covers all columns selected by the query.

**Comment by iswarezwp** [comment]
@robertpang should i build from source to try this feature?

**Comment by kmuthukk** [comment]
hi @iswarezwp : You can install from binary (say on a Mac or Linux or using docker) and try this feature out as well. Please see the quick install here: https://docs.yugabyte.com/latest/quick-start/.

If you are looking to try it out on a multi-node deployment, follow the steps here:

https://docs.yugabyte.com/latest/deploy/multi-node-cluster/

**Comment by rkarthik007** [comment]
Hi @iswarezwp - we have updated the "explore" section of the docs so you can quickly get started with secondary indexes:
https://docs.yugabyte.com/latest/explore/secondary-indexes/

Note that you need the 0.9.9 version (current latest release, or later) for this to work.

**Comment by shreklei** [comment]
@robertpang @kmuthukk thank you very much! I will test it according to your suggestion.

**Comment by ddorian** [comment]
You guys can support secondary index when partitioning by the same columns without distributed transaction.  Ex: table ((user_id),event_id) index ((user_id),timestamp, event_id)

**Comment by rkarthik007** [comment]
@ddorian - if we did that, we would have to store the index per hash partition, and serve consistent data by doing a fan out query - which becomes less efficient as we add nodes.

For example, the second query shown below might need data from all hash partitions if we stored a per-partition index. The most efficient index organization is to create an index table that has a completely different partitioning scheme:
```
select sum(amount) from orders where order_date = '2018-04-02';
```

But we're in the early stages of working on a couple of constructs that I believe you're referring to:
- #79 ability to "co-partition" tables (issue has details)
- #25 supporting local indexes where there is a lot of data inside the partition and there is a need to locate that data efficiently

**Comment by robertpang** [comment]
Implemented in commit [e321311](https://github.com/YugaByte/yugabyte-db/commit/e32131149d9208eab29645ac3bfda0e302e4c2a2)
